pkg/endpoint/bpf.go

```diff
@@ -1204,36 +1270,48 @@ func (e *Endpoint) applyPolicyMapChanges() error {
 		}
 	}
 
-	// Add policy map entries before deleting to avoid transient drops
-	for keyToAdd := range adds {
-		// AddVisibilityKeys() records changed keys in both 'deletes' (old value) and 'adds' (new value).
-		// Remove the key from 'deletes' to keep the new entry.
-		delete(deletes, keyToAdd)
+	// If the lockdown mode is enabled and the desiredPolicy length
+	// exceeds the bpf map entry limit, *and* the endpoint is in a
+	// ready state.
+	if option.Config.EnablePolicyOverflowEndpointLockdown &&
+		e.desiredPolicy.GetPolicyMap().Len() > policymap.MaxEntries &&
+		e.setState(StateLockingDown, "Desired policy state exceeds map length maximum") {
+		e.getLogger().WithFields(logrus.Fields{
+			logfields.EndpointID: e.ID,
+		}).Warn("Policy Map exceeds the policy map max entries limit, locking endpoint down...")
 
-		entry, exists := e.desiredPolicy.GetPolicyMap().Get(keyToAdd)
-		if !exists {
+		err := e.endpointPolicyLockdown()
+		if err != nil {
 			e.getLogger().WithFields(logrus.Fields{
-				logfields.AddedPolicyID: keyToAdd,
-			}).Warn("Tried adding policy map key not in policy")
-			continue
-		}
-
-		// Redirect entries currently come in with a dummy redirect port ("1"), replace it with
-		// the actual proxy port number, or with 0 if the redirect does not exist yet. This is
-		// due to the fact that proxies may not yet have bound to a specific port when a proxy
-		// policy is first instantiated.
-		if entry.IsRedirectEntry() {
-			entry.ProxyPort = e.realizedRedirects[policy.ProxyIDFromKey(e.ID, keyToAdd)]
+				logfields.EndpointID: e.ID,
+			}).WithError(err).Error("Failed to lockdown endpoint. The endpoint may be compromised, consider quarantining or shutting down this node.")
```

Comment:
When this occurs, I'm assuming that the endpoint will be stuck in the new "locking-down" state? If an endpoint is stuck in that state, then it is considered "compromised" because Cilium cannot enforce lockdown mode on it? I think if that's the implication, then it needs to be clearer. The docs also need to be updated with this information, along with the new states.

Reply:
> I'm assuming that the endpoint will be stuck in the new "locking-down" state?

No. It is not stuck. Other reconciliation attempts can bring it out of this state, including the policy map being relieved of pressure.

> The docs also need to be updated with this information, along with the new states.

Fair enough.